### Ensembled_Technique & Random Forest

There are 3 Ensembled Technique 

1.Bagging

2.Boosting

3.Stacking

#### Bagging ~ Bootstrap Aggregation
It Create Random Sample of Datapoint(Random Train Set) with Sampling with Replacement(A Particular Datapoints can be repeated in multiple Random sample)

Independent of Previous Model (Parallel in nature)

prediction Average / mode is taken

In [1]:
# Bagged Decision Trees for Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import preprocessing

#Load the data set
cancer = pd.read_csv("C:/Users/Akaash/Downloads/cancer.csv")

#Complete Cancer dataset - applying lable encoder to species column - (Y Variable is Categorial)
label_encoder = preprocessing.LabelEncoder()
cancer['diagnosis'] = label_encoder.fit_transform(cancer['diagnosis']) 

# Dividing our data into input and output variables 
X = cancer.iloc[:,1:]
Y = cancer.iloc[:,0]

# Creating K-Fold Instance
seed = 7
kfold = KFold(n_splits=10,random_state=seed,shuffle = True)

#Creating the Instance of DecisionTreeClassifier()
classifier = DecisionTreeClassifier()

#Training the Model -- Bagging
num_trees = 100
#base_estimator= instance of the Technique used which is either DTclassifier or DTRegressor
model = BaggingClassifier(base_estimator=classifier,n_estimators=num_trees,random_state=seed) 

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)


In [2]:
# Accuracy for Each K-Fold
results

array([0.96491228, 1.        , 0.94736842, 0.94736842, 0.96491228,
       0.94736842, 0.96491228, 0.96491228, 0.98245614, 0.98214286])

In [3]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

96.66353383458647

Inference: Using the Bagging Ensembled Technique the Final Accuracy is 96.66%

#### Random Forest
It is Based on the Concept of Bagging where Datapoint are Randomly Sampled.

It Splits Data Point as well as the features Randomly and Creates Decision Tree

In [4]:
# Random Forest Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn import preprocessing

#Load the data set
cancer = pd.read_csv("C:/Users/Akaash/Downloads/cancer.csv")

#Complete Cancer dataset - applying lable encoder to species column - (Y Variable is Categorial)
label_encoder = preprocessing.LabelEncoder()
cancer['diagnosis'] = label_encoder.fit_transform(cancer['diagnosis']) 

# Dividing our data into input and output variables 
X = cancer.iloc[:,1:]
Y = cancer.iloc[:,0]

# Creating K-Fold Instance
kfold = KFold(n_splits=10, random_state=7,shuffle = True)

#Training the Model ~ Random Forest 
num_trees = 100
max_features = 3 
model = RandomForestClassifier(n_estimators=num_trees, max_features=max_features) #max_features will sample randomly 3 Features

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)

In [5]:
# Accuracy for Each K-Fold
results

array([0.94736842, 0.98245614, 0.96491228, 0.96491228, 0.94736842,
       0.92982456, 0.98245614, 0.94736842, 0.98245614, 0.96428571])

In [6]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

96.13408521303256

Inference: Using The Random Forest the Final Accuracy is 96.13%

#### Boosting ~ AdaBoost
Stage Weight of That Datapoint Who give Error is updated in the next Training Set of Next model, and the Next Model Try to minimize the Error / Boost the Performance of the model

It is Dependent on the Previous model i.e sequential in Nature

In [7]:
# AdaBoost Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import AdaBoostClassifier
from sklearn import preprocessing

#Load the data set
cancer = pd.read_csv("C:/Users/Akaash/Downloads/cancer.csv")

#Complete Cancer dataset - applying lable encoder to species column - (Y Variable is Categorial)
label_encoder = preprocessing.LabelEncoder()
cancer['diagnosis'] = label_encoder.fit_transform(cancer['diagnosis']) 

# Dividing our data into input and output variables 
X = cancer.iloc[:,1:]
Y = cancer.iloc[:,0]

# Creating K-Fold Instance
seed=7
kfold = KFold(n_splits=10, random_state=seed,shuffle = True)

#Training the Model ~ Adaboost (Boosting)
num_trees = 12
model = AdaBoostClassifier(n_estimators=num_trees, random_state=seed)

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)


In [8]:
# Accuracy for Each K-Fold
results

array([0.96491228, 0.98245614, 0.96491228, 0.96491228, 0.92982456,
       0.92982456, 0.9122807 , 0.92982456, 0.96491228, 0.96428571])

In [9]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

95.08145363408522

Inference: Using the Boosting Ensembled Technique the Final Accuracy is 95.08%

#### Stacking

Through Stacking Can Use Multiple Algorithms At Once the TAle mean/Mode of the Prediction

In [10]:
# Stacking Ensemble for Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn import preprocessing

import warnings
warnings.filterwarnings("ignore")  #--to ignore warnings

#Load the data set
cancer = pd.read_csv("C:/Users/Akaash/Downloads/cancer.csv")

#Complete Cancer dataset - applying lable encoder to species column - (Y Variable is Categorial)
label_encoder = preprocessing.LabelEncoder()
cancer['diagnosis'] = label_encoder.fit_transform(cancer['diagnosis']) 

# Dividing our data into input and output variables 
X = cancer.iloc[:,1:]
Y = cancer.iloc[:,0]

# Creating K-Fold Instance
seed=7
kfold = KFold(n_splits=10, random_state=seed,shuffle = True)

# create the sub models - 3 Different Models
estimators = []
model1 = LogisticRegression(max_iter=500)
estimators.append(('logistic', model1))
model2 = DecisionTreeClassifier()
estimators.append(('cart', model2))
model3 = SVC()
estimators.append(('svm', model3))

# create the ensemble model
ensemble = VotingClassifier(estimators)

#Predicting & Getting the Model Score with kfold
results = cross_val_score(ensemble, X, Y, cv=kfold)


In [11]:
# Accuracy for Each K-Fold
results

array([0.9122807 , 0.98245614, 0.98245614, 0.98245614, 0.94736842,
       0.9122807 , 0.98245614, 0.94736842, 0.94736842, 0.96428571])

In [12]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

95.60776942355889

Inference: Using the Stacking Ensembled Technique the Final Accuracy is 95.60%